srtcore/tsbpd_time.h

```diff
@@ -0,0 +1,155 @@
+class CTsbpdTime
+{
+    typedef srt::sync::steady_clock  steady_clock;
+    typedef steady_clock::time_point time_point;
+    typedef steady_clock::duration   duration;
+    typedef srt::sync::Mutex         Mutex;
+
+public:
+    CTsbpdTime()
+        : m_bTsbPdMode(false)
+        , m_tdTsbPdDelay(0)
+        , m_bTsbPdWrapCheck(false)
+    {
+    }
+
+    /// Set TimeStamp-Based Packet Delivery Mode (receiver).
+    /// @param [in] timebase localtime base (uSec) of packet time stamps including buffering delay.
+    /// @param [in] wrap Is in wrapping period.
+    /// @param [in] delay aggreed TsbPD delay (buffering latency).
+    void setTsbPdMode(const time_point& timebase, bool wrap, duration delay);
+
+    /// @brief Check if TSBPD logic is enabled.
+    /// @return true if TSBPD is enabled.
+    bool isEnabled() const { return m_bTsbPdMode; }
+
+    /// @brief Apply new state derived from other members of a socket group.
+    /// @param timebase TSBPD base time.
+    /// @param wrp wrap period (enabled or not).
+    /// @param delay TSBPD delay.
+    /// @param udrift Clock drift.
+    void applyGroupTime(const time_point& timebase, bool wrp, uint32_t delay, const duration& udrift);
+
+    /// @brief Apply new clock state (TSBPD base and drift) derived from other members of a socket group.
+    /// @param timebase TSBPD base time.
+    /// @param wrp the state of wrapping period (enabled or disabled).
+    /// @param udrift Clock drift.
+    void applyGroupDrift(const time_point& timebase, bool wrp, const duration& udrift);
+
+    /// @brief Add new drift sample from an ACK-ACKACK pair.
+    /// ACKACK pakcets are sent immideatly (except for UDP buffering).
+    /// 
+    /// @param [in] pktTimestamp Timestamp of the arrived ACKACK packet.
+    /// @param [out] w_udrift Current clock drift value
+    /// @param [out] w_newtimebase Current TSBPD base time
+    /// 
+    /// @return true if TSBPD base time was changed, false otherwise.
+    bool addDriftSample(uint32_t                  pktTimestamp,
+                        steady_clock::duration&   w_udrift,
+                        steady_clock::time_point& w_newtimebase);
+
+    /// @brief Get packet TSBPD time with buffering delay.
+    /// The target time when to deliver the packet to an upstream application.
+    /// @param [in] usPktTimestamp 32-bit value of packet timestamp field (microseconds).
+    ///
+    /// @returns Packet TSBPD base time with buffering delay.
+    time_point getPktTsbPdTime(uint32_t usPktTimestamp) const;
+
+    /// @brief Get packet TSBPD base time without buffering delay.
+    /// @param [in] usPktTimestamp 32-bit value of packet timestamp field (microseconds).
+    ///
+    /// @returns Packet TSBPD base time without buffering delay.
+    time_point getPktTsbPdBaseTime(uint32_t usPktTimestamp) const;
+
+    /// @brief Get TSBPD base time considering possible carryover
+    /// when packet timestamp is overflown and continues from 0.
```

Comment:
Not sure what you mean here. Does this function get the TSBPD base time value in order to compare with the timestamp for a packet to see if it has exceeded the wrap value (and therefore should be re-evaluated to continue from 0 as it should)?

Also, is this inversion in function names intentional?
  time_point getPktTsbPd**BaseTime**(uint32_t usPktTimestamp) const;
  time_point getTsbPd**TimeBase**(uint32_t usPktTimestamp) const;

Reply:
@stevomatthews  This naming is something that also looks confusing to me and I would love to improve,

Let us say that Packet TSBPD time is the time when a packet can be extracted from the receiver buffer.
Essentially, it consists of `usPktTsbpdTime = usTsbpdTimeBase + usPktTimeStamp + usTsbpdDelay + usDrift`,
where `usTsbpdTimeBase` takes possible packet timestamp carryover into account.

Then:
- `getTsbPdTimeBase(uint32_t usPktTimestamp)` returns `usTsbpdTimeBase` for a packet (with possible carryover).
- `getPktTsbPdBaseTime(uint32_t usPktTimestamp)` returns `usTsbpdTimeBase + usPktTimeStamp`.
- `getPktTsbPdTime(uint32_t usPktTimestamp)` returns `usTsbpdTimeBase + usPktTimeStamp + usTsbpdDelay + usDrift`.